### PR TITLE
Moving towards 1.1.0

### LINKED ISSUE #2: Handle attachements (multipart) using PGP/MIME

### LINKED ISSUE #3: Use setuptools for easy install?
See also #1

### LINKED ISSUE #5: sign outgoing Mails
Hi,

first of all, thanks for the script, it's working great :)

Would it be possible to not only crypt outgoing mails but also sign them? I tried by myself to implement that feature, but unfortunately I've no clue about python.

### LINKED ISSUE #12: Add support for Python 3.4
https://travis-ci.org/infertux/zeyple/jobs/68632995

### PR DESCRIPTION
**Pull request by acatton**
I don't use this project, but I'll need to use in the future. And you seem to have a userbase.

This is my ongoing work to move towards 1.1.0.

- [x] Clean up the overuse of "pseudo"-private methods
- [x] Add the ability to override the GPG executable path
- [x] Don't modify `os.environ`
- [x] Get rid of `assert is_encrypted(...)` in the tests, use `assert self.decrypt(encrypted_content) == content`
- [x] Use a temporary directory as GPG `--homedir` during the tests
- [x] Fix #2 
- [ ] Fix #12 
- [ ] Fix #5 
- [x] ~~Fix #3~~ See my comment on #3 
- [x] Include #10

I don't have a setup with zeyple, can someone test my fixes on their setup?

### CODE REVIEW

**Comment by infertux** [comment]
Awesome work @acatton! I think we should leave #5 out of 1.1.0 though, this is already a lot of changes.

File: tests/test_zeyple.py [hunk 1]
```diff
@@ -106,10 +141,32 @@ def test_process_message_with_multipart_message(self):
...
+
+            """ + content + """
+        """), [TEST1_EMAIL, TEST2_EMAIL])
+
+        assert len(emails) == 2  # It had two recipients
 
+        for m in emails:
+            assert m['X-Zeyple'] is not None
+            payload = self.decrypt(m.get_payload()).strip()
+            assert payload == six.b(content)
```
Comment (infertux) [inline tests/test_zeyple.py]: I think it'd be great to replace `assert is_encrypted(...)` with `assert self.decrypt(...) == plain_text_content` for all tests now that we have a way to decrypt messages. `is_encrypted()` was to make sure the email was actually encrypted but if we could both check that it is encrypted and properly decryptable, it's even better. Thoughts?

File: zeyple/zeyple.py [hunk 2]
```diff
@@ -5,8 +5,10 @@
 import os
 import logging
 import email
+import email.mime.multipart
 import smtplib
 import gpgme
+import copy
 from io import BytesIO
 try:
     from configparser import SafeConfigParser  # Python 3
```
Comment (infertux) [inline zeyple/zeyple.py]: Not part of your changes but shouldn't we use `from six.moves.configparser import ConfigParser` instead here?

File: zeyple/zeyple.py [hunk 3]
```diff
@@ -24,24 +26,39 @@
...
-    def process_message(self, message, recipients):
+        if self.config.has_option('gpg', 'executable'):
+            executable = self.config.get('gpg', 'executable')
+        else:
+            executable = None  # Default value
+
+        home_dir = self.config.get('gpg', 'home')
+
+        ctx = gpgme.Context()
+        ctx.set_engine_info(protocol, executable, home_dir)
```
Comment (infertux) [inline zeyple/zeyple.py]: Nice! I remember having trouble with that and had to resort injecting the `GNUPGHOME` env variable which is ugly.

File: zeyple/zeyple.py [hunk 4]
```diff
@@ -24,24 +26,39 @@
...
+        return ctx
+
+    def process_message(self, message_data, recipients):
         """Encrypts the message with recipient keys"""
 
-        message = email.message_from_string(message)
-        logging.info("Processing outgoing message %s", message['Message-id'])
+        in_message = email.message_from_string(message_data)
+        logging.info(
+            "Processing outgoing message %s", in_message['Message-id'])
```
Comment (infertux) [inline zeyple/zeyple.py]: That indentation looks confusing to me, is that the way you're supposed to indent in proper Python?

File: zeyple/zeyple.py [hunk 5]
```diff
@@ -81,27 +105,30 @@ def _send_message(self, message, recipient):
...
+        smtp = smtplib.SMTP(self.config.get('relay', 'host'),
+                            self.config.get('relay', 'port'))
 
         smtp.sendmail(message['From'], recipient, message.as_string())
         smtp.quit()
 
         logging.info("Message %s sent", message['Message-id'])
 
-    def _load_configuration(self, filename='zeyple.conf'):
+    def load_configuration(self, fname):
```
Comment (infertux) [inline zeyple/zeyple.py]: Not a big fan of using abbreviations in programming in general. Without context, it's hard to figure out what is `fname` - is it a _function name_, a _file name_, a _first name_, etc.? And most importantly, when you want to use it, you have to ask yourself _"Did I abbreviate this variable or not? Oh right, I did, but was it `fname` or `fn` or maybe something else?"_.

**Comment by infertux** [comment]
This is great @acatton! I left some nitpicking comments on the diff.

I'll test your changes on my setup and let you know if everything works fine.

**Comment by infertux** [comment]
Just like https://github.com/infertux/zeyple/issues/2#issuecomment-126308795, plain text emails work fine but multipart emails can't be decrypted.

---

Actual result:

```
Content-Type: multipart/mixed; boundary="BOKacYhQ+x31HxR3"
Content-Disposition: inline

-----BEGIN PGP MESSAGE-----
Version: GnuPG v1.4.12 (GNU/Linux)

[...]
```

I can decrypt `[...]` on the command line to:

```
Content-Type: multipart/mixed; boundary="===============6531299818164456294=="
MIME-Version: 1.0

--===============6531299818164456294==
Content-Type: text/plain; charset=us-ascii
Content-Disposition: inline


--===============6531299818164456294==
Content-Type: image/png
Content-Disposition: attachment; filename="image.png"
Content-Transfer-Encoding: base64

[base64 image]
```

---

Expected result using PGP/MIME (expected by Thunderbird at least):

```
Content-Type: multipart/encrypted;
 protocol="application/pgp-encrypted";
 boundary="tEksdshbSOMXJQgTtgeuSwKpLVMQ9INAl"

This is an OpenPGP/MIME encrypted message (RFC 4880 and 3156)
--tEksdshbSOMXJQgTtgeuSwKpLVMQ9INAl
Content-Type: application/pgp-encrypted
Content-Description: PGP/MIME version identification

Version: 1

--tEksdshbSOMXJQgTtgeuSwKpLVMQ9INAl
Content-Type: application/octet-stream; name="encrypted.asc"
Content-Description: OpenPGP encrypted message
Content-Disposition: inline; filename="encrypted.asc"

-----BEGIN PGP MESSAGE-----

[...]
```

`[...]` decrypts to:

```
Content-Type: multipart/mixed;
 boundary="------------020007060702010403070805"

This is a multi-part message in MIME format.
--------------020007060702010403070805
Content-Type: text/plain; charset=utf-8
Content-Transfer-Encoding: quoted-printable



--------------020007060702010403070805
Content-Type: image/png; name="image.png"
Content-Transfer-Encoding: base64
Content-Disposition: attachment; filename="image.png"

[base64 image]
```

Basically, it looks like you need to wrap everything into a `encrypted.asc` file and use `Content-Type: multipart/encrypted` as the wrapping MIME type.

This is just my empirical findings. I haven't dug into the RFCs but I'll to it when I get a chance.

File: zeyple/zeyple.py (hunk 2 above)
Comment (acatton) [inline zeyple/zeyple.py]: It depends. Do you want to make zeyple depend on six? Or just its tests?

File: zeyple/zeyple.py (hunk 4 above)
Comment (acatton) [inline zeyple/zeyple.py]: The line is too long for 80 characters. That's the way you're supposed to break a long line. [Hanging indents *may* be indented to other than 4 spaces.](https://www.python.org/dev/peps/pep-0008/#indentation).

File: zeyple/zeyple.py (hunk 5 above)
Comment (acatton) [inline zeyple/zeyple.py]: Yeah, with the hindsight, fname was bad. I was trying to follow [Mecurial's naming conventions](https://mercurial.selenic.com/wiki/CodingStyle#Naming_conventions) but it looks bad actually.

**Comment by acatton** [comment]
@infertux this is great. The multipart encryption is what I wanted you to test, and that what I need to work on then. And thank you @Nithanim for all the debugging information, this is **very** useful.

I'll try to address all the reported issues I soon as I can.

**Comment by acatton** [comment]
@infertux regarding #5, I think it's pretty important feature. I want to be able to generate keys for my servers, and make sure my servers actually sent these emails, and that nobody is spoofing them.

Moreover, I think it should be pretty straightforward to implement.

File: tests/test_zeyple.py (hunk 1 above)
Comment (acatton) [inline tests/test_zeyple.py]: Right know, `self.decrypt` doesn't play well with non-ascii characters, and I haven't tried to figure it out yet. This is why the test I wrote use ascii characters for now. But yeah, that sounds like the right way to go. I added it to the todolist of this pull request.

**Comment by acatton** [comment]
The tests should be totally broken right now. I need to fix them before merging, but this should behave according to the RFCs. Can someone try to send emails with and without attachment?

@infertux @Nithanim.

Thanks.

**Comment by acatton** [comment]
@infertux you are right. I think we should move #12 and #3 to another version. What do you think?

**Comment by Nithanim** [comment]
@acatton I tested your latest commit https://github.com/acatton/zeyple/commit/1720c0249763fb5f90227f477eb230551cc5e4e2 with mixed success.
Plaintext without attachment is broken now and won't display anything at all (although it is stating that it decrypted something). Plaintext with attachment works great! Html only without attachment is displayed somewhat plain only and then in html again after an <hr> but this is acceptable in some way but not perfect. Html with attachment on the other side looks now exactly as it should again.
You did an fantastic job there!

Here are some test-files again: https://www.dropbox.com/sh/awbjoff0naycmv3/AAB3SLvtLbSO_7uGpl81T3hva/2nd?dl=0%3Flst (link is the same as last time).

**Comment by acatton** [comment]
@Nithanim this is definitely not mergeable as is then. It supposed to work perfectly.

Thank you for the feedback, i'll be fixing that when I have time :) .

**Comment by Nithanim** [comment]
@acatton I don't know how useful it will be but I am trying to build a vagrant box to have a easy-to-setup test environment. It looks good so far. However it currently lacks the enigmail setup and the gpg keys.
It is very likely to have some problems. For example if you are on windows you need to "git config core.eol lf && git checkout-index --force --all" in a clean workingdir to workaround a python(?) bug in the guest system.
For the fearless: https://github.com/Nithanim/zeyple/tree/vagrant

PS: Got my first external html mail and it was successfully encrypted by zeyple! Exactly like in my simple test. That is so awesome!

**Comment by infertux** [comment]
Sorry for the delay @acatton. I tried https://github.com/acatton/zeyple/commit/1720c0249763fb5f90227f477eb230551cc5e4e2 and I'm experiencing the same issue described by @Nithanim.

I compared it with properly encrypted plain text emails and if I understand that right, you need to convert the plain text into a MIME payload [here](https://github.com/acatton/zeyple/blob/1720c0249763fb5f90227f477eb230551cc5e4e2/zeyple/zeyple.py#L106) - maybe with [this](https://docs.python.org/2/library/email.mime.html#email.mime.text.MIMEText) or [that](https://docs.python.org/2/library/email.mime.html#email.mime.nonmultipart.MIMENonMultipart)?

In other words, `payload` shouldn't be a mere `hello world` string but this:

```
Content-Type: multipart/mixed; boundary="MvkF3o3R8o27A6vwDL8mILdrsqfvHWfQC"

--MvkF3o3R8o27A6vwDL8mILdrsqfvHWfQC
Content-Type: text/plain; charset=utf-8
Content-Transfer-Encoding: quoted-printable

hello world

--MvkF3o3R8o27A6vwDL8mILdrsqfvHWfQC--```

Hope that makes sense.

**Comment by infertux** [comment]
Regarding the _1.1.0_ release @acatton:

- let's put off #12 to a future releae since those are pretty recent Python versions anyway,
- if you think #5 is pretty straightforward to implement, then I'm okay to include it in _1.1.0_ otherwise I think we should do it later since this release is already quite big in term of changes.

**Comment by infertux** [comment]
Relevant parts of the RFCs about the encrypted payload: http://tools.ietf.org/html/rfc3156#section-6.1 & http://tools.ietf.org/html/rfc1847#section-2.2

> The contents of the body part to be protected is prepared according
        to a local convention.  **The contents are then transformed into a
        MIME body part in canonical MIME format, including an appropriate
        set of MIME headers.** [emphasis mine]

This seems to apply to any `Content-Type`, including `text/plain`.

/cc @acatton

**Comment by infertux** [comment]
Alright, I updated the tests to reflect what is expected by the RFCs in [this branch](https://github.com/infertux/zeyple/tree/acatton/moving-towards-1.1.0-infertux). There is [one failing test](https://travis-ci.org/infertux/zeyple/jobs/76127726). If we manage to fix it without breaking other tests, it should encrypt everything properly :smiley:

**Comment by infertux** [comment]
Just gave you push access to the repo @acatton so we can both work on the same branch. I opened a new PR with branch `moving-towards-1.1.0` at #17.
